File: leetcode/medium/347-topKFrequentElements.js
// Given a non-empty array of integers, return the k most frequent elements.

// Example 1:
// Input: nums = [1,1,1,2,2,3], k = 2
// Output: [1,2]

// Example 2:
// Input: nums = [1], k = 1
// Output: [1]

// first pass solution
var topKFrequent = function(nums, k) {
    let charCount = {};
    let result = [];
    
    for (let num of nums) {
        if (!(num in charCount)) charCount[num] = 0;
        charCount[num] += 1;
    }
    
    let entries = Object.entries(charCount).sort((a, b) => b[1] - a[1]);
    
    for (let i = 0; i < k; i++) {
        result.push(entries[i][0]);
    }
    return result;
};

File: leetcode/medium/347-topKFrequentElements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> hashMap = new HashMap<>();
        int[] output = new int[k];
        Queue<Integer> heap = new PriorityQueue<>(
            (n1, n2) -> hashMap.get(n1) - hashMap.get(n2));
        
        for (int n : nums) {
            if (hashMap.containsKey(n)) {
                hashMap.put(n, hashMap.get(n) + 1);
            } else {
                hashMap.put(n, 1);
            }
        }
        
        for (int key : hashMap.keySet()) {
            heap.add(key);
            if (heap.size() > k) heap.poll();
        }
        
        for (int i = k - 1; i >= 0; i--) {
            output[i] = heap.poll();
        }
        
        return output;
    }
}
